### PR TITLE
[MM-42402]: compact post editing alignment fix

### PR DESCRIPTION
**Pull request by michelengelen**
#### Summary
placeholder text and emoji button should now align and the wonky behavior when replacing the post body with the edit input is now gone as well (it got somehow rendered in flow and when animation finished it got ripped out of the flow)

#### Ticket Link
[MM-42402](https://mattermost.atlassian.net/browse/MM-42402)

#### Screenshots
|  before  |  after  |
|----|----|
|Screenshot 2022-03-10 at 14 32 40|Screenshot 2022-03-10 at 14 32 29|

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Review by devinbinnie — APPROVED** [review]

**Review by matthewbirtch — APPROVED** [review]
Looks good @michelengelen

The thread footer has some misalignment in compact mode, but I've created a separate ticket to fix that: https://mattermost.atlassian.net/browse/MM-42472

**Comment by jgilliam17** [comment]
Looks good @michelengelen 
Placeholder text and emoji icon are aligned.
Seeing one issue when saving edited post. Text indents to the right for a moment before it saves justified.

https://user-images.githubusercontent.com/52937121/157986660-7d0fd8bf-da07-4760-b47b-8321e7b8f7ec.mov

**Comment by matthewbirtch** [comment]
> ok, It took my a while to figure that out, but we might have a problem with the compact format.
> 
> to calculate and animate the height on the post body it is necessary to set the `overflow` value to `hidden`. The problem is the post header. It is aligned via `float: left` to let it break out of the normal element flow. `overflow: hidden` on the other hand prevents the content of the element to flow around floating elements. I did create a simple showcase on codepen where you can see the behavior of `overflow: hidden` in combination with floating elements
> 
> https://codepen.io/michelengelen/pen/KKZdmwj
> 
> The big question is where do we want to go with this. The underlying problem is (again) the structure of the elements, so we could go deep down the rabbit hole and refactor all of it, as with #9482 or we can find a way around it. Both options include a lot of work, since a workaround would mean building our own height calculation component around it. Thats probably less impactful on existing code, but still a rather heavy task.
> 
> WDYT @matthewbirtch ?
> 
> /cc @hmhealey

@michelengelen is this in reference to the last issue that @jgilliam17 identified?

I'm not really qualified to suggest how to resolve the bug, so I don't really want to suggest one way or the other. From a UX perspective this definitely feels broken and buggy to me. I'm certain that a much smaller percentage of users have compact mode enabled, so fixing this bug will have much less impact, BUT I do think we should find a way to resolve it eventually. I would be okay with it being a separate follow-up though.

**Comment by michelengelen** [comment]
@matthewbirtch Well ... i could not let it be like that! :D
That's why i have already build a fix, but I need to fix some tests first before I will push it ... another good news: I could remove the package I originally added since I build my own height changing mechanism now!

**Comment by matthewbirtch** [comment]
> @matthewbirtch Well ... i could not let it be like that! :D That's why i have already build a fix, but I need to fix some tests first before I will push it ... another good news: I could remove the package I originally added since I build my own height changing mechanism now!

Ok, well that's great. It didn't sound like you already had a solution! Thanks for tackling it and I appreciate the high standard of quality!

**Comment by michelengelen** [comment]
@devinbinnie and @pvev could you please go over this once more? I have changed a substantial part of it to fix the "jumpy" compact display bug and would like to have at least one more dev review. 🙇🏼

**Comment by pvev** [comment]
> @devinbinnie and @pvev could you please go over this once more? I have changed a substantial part of it to fix the "jumpy" compact display bug and would like to have at least one more dev review. 🙇🏼

sure!

**Comment by michelengelen** [comment]
@pvev and @devinbinnie this is now ready ... I had an issue with textbox alignment while transitioning in, but that is now fixed and should even be more performant and smooth with this last commit! Thanks for bearing with me for so long! 🙇🏼

File: components/post_view/post/post.jsx [hunk 1]
```diff
@@ -412,6 +417,50 @@ export default class Post extends React.PureComponent {
...
+                compactDisplay={this.props.compactDisplay}
+                isCommentMention={this.props.isCommentMention}
+                isFirstReply={this.props.isFirstReply}
+                handleFileDropdownOpened={this.handleFileDropdownOpened}
+            />
+        );
+
+        const threadFooter = isCollapsedThreadsEnabled && !post.root_id && (hasReplies || post.is_following) ? <ThreadFooter threadId={post.id}/> : null;
+
+        const showSlot = isBeingEdited ? 2 : 1;
```
Comment (pvev) [inline components/post_view/post/post.jsx:447]: maybe set this numeric values in constant with meaningful names?

File: components/common/auto_height_switcher.tsx [hunk 2]
```diff
@@ -47,32 +48,55 @@ const AutoHeightSwitcher = ({showSlot, slot1, slot2, duration = 250, shouldScrol
...
+    // necessary to override this on element-level since a rule from _post.scss has a higher specificity
+    // and would override the display property
+    const fixedStyles: CSSProperties = {
+        transitionProperty: 'height',
+        transitionDuration: `${duration}ms`,
+        transitionTimingFunction: 'ease',
+    };
+
+    /**
+     * the lifecycle of this transition-component is divided in 3 different stages:
```
Comment (pvev) [inline components/common/auto_height_switcher.tsx:60]: nice comments

File: components/post_view/post/post.jsx (hunk 1 above)
Comment (michelengelen) [reply components/post_view/post/post.jsx]: done!

**Comment by michelengelen** [comment]
@jgilliam17 this is now ready for another round of QA review

**Comment by jgilliam17** [comment]
Thanks @michelengelen 
Tested, looks good to merge.
- Verified edit box item alignment and text when transitioning to saved post.
-[ E2E report](https://automation-dashboard.vercel.app/cycles/6edbe114-ba84-47ec-bf03-d1236888dcf7) looks good too, no PR related failures

**Review by jgilliam17 — APPROVED** [review]

**Pull request merged**